dvc/api.py

```diff
@@ -33,21 +34,154 @@ def get_url(path, repo=None, rev=None, remote=None):
         return dvc_repo.cloud.get_url_for(remote, checksum=md5)
 
 
+class _OpenContextManager(GCM):
+    def __init__(
+        self, func, args, kwds
+    ):  # pylint: disable=super-init-not-called
+        self.gen = func(*args, **kwds)
+        self.func, self.args, self.kwds = func, args, kwds
+
+    def __getattr__(self, name):
+        raise AttributeError(
+            "dvc.api.open() should be used in a with statement."
+        )
+
+
 def open(  # noqa, pylint: disable=redefined-builtin
-    path, repo=None, rev=None, remote=None, mode="r", encoding=None
+    path: str,
+    repo: Optional[str] = None,
+    rev: Optional[str] = None,
+    remote: Optional[str] = None,
+    mode: str = "r",
+    encoding: Optional[str] = None,
 ):
     """
-    Open file in the supplied path tracked in a repo (both DVC projects and
-    plain Git repos are supported). For Git repos, HEAD is used unless a rev
-    argument is supplied. The default remote is tried unless a remote argument
-    is supplied. It may only be used as a context manager:
```

Comment:
Is the PR title accurate here or even the primary purpose of the PR  😄? I like what you are doing, but I think the existing first paragraph is a nice summary. Not sure if we need to amend the docs, but feels like we are losing some useful info at the top (supporting DVC or Git repos, use HEAD and default remote by default).

Reply:
I don't have a problem with keeping the docstring changes in this PR as long as the general doc changes are kept in a separate commit from the mode=write behavior related changes

Reply:
@dberenbaum Do you think that the info about argument defaults should be present in the "short description" section of the docstring?

Reply:
> but what about (both DVC projects and plain Git repos are supported)?

This should be documented in the `repo` argument. 

I extracted the content from (https://dvc.org/doc/api-reference/open):

```
        repo (str, optional): location of the DVC project.
            Defaults to the current project (found by walking up from the
            current working directory tree).
            It can be a URL or a file system path.
            Both HTTP and SSH protocols are supported for online Git repos
            (e.g. [user@]server:project.git).
```

Do you think something is missing?

Reply:
I don't think it's that clear that it can be used for either DVC repos or regular Git repos, but I'm not sure anyone is using it for that anyway and definitely don't want to keep holding up getting this merged, so let's merge it.

Reply:
> I don't think it's that clear that it can be used for either DVC repos or regular Git repos

Updated the description